sympy/stats/drv.py

```diff
@@ -157,3 +165,45 @@ def compute_characteristic_function(self, expr, **kwargs):
             return self.distribution.compute_characteristic_function(**kwargs)
         else:
             raise NotImplementedError()
+
+    def restricted_domain(self, condition):
+        rvs = random_symbols(condition)
+        assert all(r.symbol in self.symbols for r in rvs)
+        if (len(rvs) > 1):
+            raise NotImplementedError(filldedent('''Multivariate discrete
+            random variables are not yet supported.'''))
+        conditional_domain = reduce_rational_inequalities_wrap(condition,
+            rvs[0])
+        conditional_domain = conditional_domain.intersect(self.domain.set)
+        return conditional_domain
+
+    def probability(self, condition):
+        _domain = self.restricted_domain(condition)
+        if condition == False or _domain is S.EmptySet:
+            return S.Zero
+        if condition == True or _domain == self.set:
+            return S.One
+        try:
+            return self.eval_prob(_domain)
+        except NotImplementedError:
+            return Probability(condition)
+
+    def eval_prob(self, _domain):
+        if isinstance(_domain, Range):
+            n = symbols('n')
+            inf, sup, step = (r for r in _domain.args)
+            summand = ((self.pdf).replace(
+                self.symbol, inf + n*step))
+            rv = summation(summand,
+                (n, 0, floor((sup - inf)/step - 1))).doit()
+            return rv
+        elif isinstance(_domain, FiniteSet):
+            pdf = Lambda(self.symbol, self.pdf)
+            rv = sum(pdf(x) for x in _domain)
+            return rv
+        elif isinstance(_domain, Union):
+            rv = sum(self.eval_prob(x) for x in _domain.args)
+            return rv
+        else:
+            raise NotImplementedError(filldedent('''Probability for
```

Comment:
we actually have a class for unevaluated probability, it's called `Probability`.

Reply:
The class is being used on L189. The unevaluated object is returned if this function raises a `NotImplemented` error at this point. Since the exception raised at this point is always caught, the error message is never actually displayed.

Reply:
Oh, I see. I would recommend anyway to try to avoid using exception handling, as it makes it difficult to debug the code. Every time you raise and catch an exception you lose the stack trace. Also the readability of the code is worsened by raising exceptions.

Reply:
Thanks for explaining. Should I remove the part where `eval_prob` raises the exception, and modify `probability` so that it returns the unevaluated object in case the return value from `eval_prob` is `None`?

Reply:
I have opened #14604 for this, including one other change.